### PR TITLE
AuditService

### PR DESCRIPTION
**Pull request by Hinton**
Companion PR to https://github.com/bitwarden/browser/pull/526.

### CODE REVIEW

File: src/services/crypto.service.ts [hunk 1]
```diff
@@ -452,6 +452,12 @@ export class CryptoService implements CryptoServiceAbstraction {
         return b64DecValue;
     }
 
+    sha1(password: string) {
```
Comment (kspearrin) [inline src/services/crypto.service.ts]: I'd rather use web crypto for this to further reduce our dependency on Forge. Web crypto will turn this into a promise-based return. https://github.com/diafygi/webcrypto-examples#sha-1---digest

File: src/services/audit.service.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    constructor(private cryptoService: CryptoService) {
+    }
+
+    async passwordLeaked(password: string) {
+        const hash = this.cryptoService.sha1(password);
+
+        const response = await fetch(this.url + hash.substr(0, 5));
+        const leakedHashes = await response.text();
+
+        const hashEnding = hash.substr(5);
```
Comment (kspearrin) [inline src/services/audit.service.ts]: use this higher up so you dont have to substring twice

File: src/services/audit.service.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { CryptoService } from 'jslib/abstractions/crypto.service';
+
+export class AuditService {
+
+    private url = 'https://api.pwnedpasswords.com/range/';
```
Comment (kspearrin) [inline src/services/audit.service.ts]: Should this be a constant outside the class?

File: src/services/audit.service.ts (hunk 2 above)
Comment (Hinton) [reply src/services/audit.service.ts]: The `substr` calls affect different portions of the hash. The first call returns the first 5 characters, while the other call returns the remaining portion of the hash.

Comment (kspearrin) [reply src/services/audit.service.ts]: I see

File: src/services/crypto.service.ts [hunk 4]
```diff
@@ -452,6 +452,17 @@ export class CryptoService implements CryptoServiceAbstraction {
         return b64DecValue;
     }
 
+    async sha1(password: string): Promise<string> {
+        const hash = await Crypto.subtle.digest(
+            {
+                name: 'SHA-1',
+            },
+            UtilsService.fromB64ToArray(window.btoa(password)),
```
Comment (kspearrin) [inline src/services/crypto.service.ts]: Why not the utf-8 to array util?

Comment (Hinton) [reply src/services/crypto.service.ts]: Whops, nice catch. Was experimenting why the result from `fromBufferToUtf8` threw errors. (Needed to convert it to Hex, not Utf8)

**Pull request merged**